sigstore/_internal/sct.py

```diff
@@ -50,41 +64,69 @@ def _pack_digitally_signed(sct: DetachedFulcioSCT, cert: Certificate) -> bytes:
     if unused:
         raise InvalidSctError(f"Unexpectedly large certificate length: {len(cert_der)}")
 
+    # No extensions are currently specified.
+    if len(sct.extension_bytes) != 0:
+        raise InvalidSctError("Unexpected trailing extension bytes")
+
+    # NOTE(ww): This is incorrect for non-embedded SCTs, since it unconditionally
+    # embeds the issuer_key_hash.
+
     # Assemble a format string with the certificate length baked in and then pack the digitally
     # signed data
-    pattern = "!BBQhBBB%ssh" % len(cert_der)
+    # fmt: off
+    pattern = "!BBQh32sBBB%ssh" % len(cert_der)
     data = struct.pack(
         pattern,
-        sct.version.value,
-        0,  # Signature Type
-        int(sct.timestamp.timestamp() * 1000),
-        0,  # Entry Type
-        len1,
-        len2,
-        len3,
-        cert_der,
-        len(sct.extensions),
+        sct.version.value,                      # sct_version
+        0,                                      # signature_type (certificate_timestamp(0))
+        int(sct.timestamp.timestamp() * 1000),  # timestamp (milliseconds)
+        sct.entry_type.value,                   # entry_type (x509_entry(0) | precert_entry(1))
+        hashlib.sha256(issuer_key).digest(),    # issuer_key_hash[32]
+        len1,                                   # \
+        len2,                                   # | opaque TBSCertificate<1..2^24-1> OR
+        len3,                                   # | opaque ASN.1Cert<1..2^24-1>
+        cert_der,                               # /
+        len(sct.extension_bytes),               # extensions (opaque CtExtensions<0..2^16-1>)
     )
```

Comment:
I'm not sure this is the problem, but I'm realizing that the serialized structure here is only the _contents_ of the digitally signed element, not the `digitally-signed struct` itself. For that, we'd need the signature and hash algorithms, plus an `opaque<1..2^16-1>` for `data` itself. 

@haydentherapper do you know if the signature is meant to be over the contents of the `digitally-signed struct`, or whether it's supposed to include the envelope as well?

Reply:
See https://github.com/google/certificate-transparency-go/blob/100db00880e6af80af0132c01b09c9eb2c4ee2dc/serialization.go#L29 - The signature is over the contents of the `digitally-signed struct`. 

I agree that the SCT is missing the signature and hash. See https://letsencrypt.org/2018/04/04/sct-encoding.html for a really great blog diving into SCT encoding

Reply:
One issue might be that the SCT doesn't contain the TBS, that's just used to generate the signature. From the RFC:

```
struct {
   Version sct_version;
   LogID id;
   uint64 timestamp;
   CtExtensions extensions;
   digitally-signed struct {
       Version sct_version;
       SignatureType signature_type = certificate_timestamp;
       uint64 timestamp;
       LogEntryType entry_type;
       select(entry_type) {
           case x509_entry: ASN.1Cert;
           case precert_entry: PreCert;
       } signed_entry;
      CtExtensions extensions;
   };
} SignedCertificateTimestamp;
```

If I understand correctly, this means that the SCT should contain only:
* A version
* Log ID
* Timestamp
* Extensions
* Signature over a version, type, timestamp, entry type, and certificate (TBS precert or ASN.1 encoded cert) - The signature containing the hash algorithm and signature

Reply:
Reading https://datatracker.ietf.org/doc/html/rfc5246#section-4.7, yea, I think this is the issue, that the SCT should be composed of some metadata and a signature, rather than metadata and the certificate. It's unclear that `digitally-signed struct` means a signature over the contents of the struct.

Reply:
Yeah, so I _think_ our current construction is right: we only build the contents of the `digitally-signed`, which looks very close to the implementation you linked above. Closing this subthread since my initial hypothesis here was wrong (it's not an envelope problem).